stdlib/option.mc

```diff
@@ -1,4 +1,132 @@
 type Option
 
 con Some : Dyn -> Option
-con None : Option
\ No newline at end of file
+con None : Option
+
+-- Returns `true` if the option is a `Some` value.
+let isNone: Option -> Bool = lam o.
+  match o with None then
+    true
+  else
+    false
+
+-- Returns `true` if the option is a `None` value.
+let isSome: Option -> Bool = lam o.
+  match o with Some _ then
+    true
+  else
+    false
+
+-- Returns `true` if the option contains a value which
+-- satisfies the specified predicate.
+let contains: Option -> (Dyn -> Bool) -> Bool = lam o. lam p.
+  match o with Some t then
+    p t
+  else
+    false
+
+-- Applies a function to the contained value (if any).
+let map: Option -> (Dyn -> Dyn) -> Option = lam o. lam f.
+  match o with Some t then
+    Some(f t)
+  else
+    None
```

Comment:
@EliasC Will MCore support function overloading? Just curious considering that there is a different map for sequences, and it would be nice to have access to both.

Reply:
That is a good question! My feeling is that MCore will _not_ support function overloading, but that Ragnar (or some other language built on top of MCore) will. For now, when there isn't even a module system to disambiguate conflicting names, we need to use a naming scheme that separates them, e.g., `optionMap` and `seqMap` (although I would very much prefer to let `map` be an alias of `seqMap` in that case).

There is also the possibility to use the same trick as in your PR for formatted printing (#25). Something like this (but with polymorphism rather than dynamic typing):

```
lang Functor
  syn FunctorWrapper =
  -- Intentionally left blank

  sem fmap (f : Dyn -> Dyn) =
  -- Intentionally left blank
end

lang SeqFunctor = Functor
  syn FunctorWrapper =
  | SeqFunctor [Dyn]

  sem fmap (f : Dyn -> Dyn) =
  | SeqFunctor s -> seqMap f s
end
```

```
utest fmap (addi 1) SeqFunctor([1,2,3]) with [2,3,4] in
...
```
We should however agree on this before we start reimplementing the entire standard library like this.

Reply:
MCore will most likely not be supporting overloading. As I see it, it should be implemented on a language on top of MCore. Ragnar (which should be the language we use for implementing the compiler in the end) should of course support it in a controlled way (most likely some variant of type classes).